### PR TITLE
[FIX] product : Improved error message for variant attribute value de…

### PR DESCRIPTION
**Pull request by fdamhaut**
…letion

What are the steps to reproduce your issue?
- Create a product with more than one attribute.
Let say color White, black and purple
- Create a 'draft' invoice for the purple product variant
- Remove the 'purple' attribute value from the product
- It will archive that variant (because the account.move linked to it)
- Try to delete the attribute value from menu Sales > Cofinfiguration > Attribute

What is the current behavior that you observe?
- technical error message

What would be your expected behavior in this case?
- non-technical message for end-users

Solution : 
- Change both message to tell user he cannot delete the value if the value has been referenced somewhere else.

opw-2623583

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by Feyensv — CHANGES_REQUESTED** [review]
1) Products are not only used on sales orders, but on a large panel of models.
Your error messages only speak of "Sales Order" but the user might not even have sale installed on his database.
2) Avoid as much as possible to try/catch all exceptions (I know there exists code like this, but it's not a good practice IMHO)
Example: we have a retry mechanism for some concurrent Access Error on the orm level, but you disable it with your `except Exception`.
3) This seems to be more an orm problem, failing to deliver a clear message in this case because he couldn't find the linked model.
Either you fix it on the orm level (good luck :D), or you make a clean solution, searching for linked products, verifying if they have linked records (method `_filter_to_unlink` on `product.product` model could be useful) and giving a cleaner explanation.

Note that you probably shouldn't give the concerned models in this explanation, unless you want to add custom code in multiple modules (not worth it for this problem IMHO).

**Comment by fdamhaut** [comment]
TY for the review @Feyensv 
Ok for 1 

I only used it because it 's indeed used a bit later in the code and after tracing all the possible links from this value, i did not found any link as once an value is unlinked from a attribute line, you cannot know if it ever was and the same apply for product_attribute as they are only linked to template and the link is removed if you "disable" the option.

Another solution would be to search env for all possible blocking record which are linked to a product.product with the given attribute value but I honestly don't think it would be a good solution anyway from a performance cost/reward point of view.
 
A simple fix would be the reason why I added a todo about archive as it would easily solve all problems as they would be archived instead of deleting but I don't know how it would affect the production database or even if it's allowed for bugfixes.
(This could also be applied to 

TL;DR - From my short experience, variant are a mess.

**Comment by fdamhaut** [comment]
Note if you believe No try/catch and unreadable error message is better than try/catch, the problem only has low impact at best.

**Comment by Feyensv** [comment]
> Another solution would be to search env for all possible blocking record which are linked to a product.product with the given attribute value but I honestly don't think it would be a good solution anyway from a performance cost/reward point of view.

That's why I was pointing out `_filter_to_unlink` which basically filters out products linked to records, which cannot be deleted.
If some products of the attribute are filtered out, it means the attribute cannot be deleted.

> A simple fix would be the reason why I added a todo about archive as it would easily solve all problems as they would be archived instead of deleting but I don't know how it would affect the production database or even if it's allowed for bugfixes. (This could also be applied to linking with all their active and past ProductTemplateAttributeValue)

Nope, new field = new column and you cannot make such a change in a stable version (so not for a bugfix), but it can indeed be a feedback for master.

> PS : regarding product, once it's unlink (I mean in the app/attribute line, not deleted), you cannot reach it from an attribute value

You mean once it's archived? By default, the ORM hides all archived records from search results, but if you really want to find it, you can make the query with `active_test=False` specified in the context (or by explicitly searching on the active field, with a domain `['|', ('active', '=', True), ('active', '=', False)]`)

> TL;DR - From my short experience, variant are a mess.

Yes, it's one of the core models used in so many places, for so long, it's become quite complex and messy with the time, and nobody dares/has the time to really clean it :D.

**Comment by fdamhaut** [comment]
> You mean once it's archived? By default, the ORM hides all archived records from search results, but if you really want to find it, you can make the query with active_test=False specified in the context (or by explicitly searching on the active field, with a domain ['|', ('active', '=', True), ('active', '=', False)])

No I mean that once you remove the possibility of such an item existing in the system (so remove the attribute value from the line, so that for example the value "Green" of attribute "Color" in product template"Tshirt" is not available anymore, you cannot have access to "Tshirt" from "Green" anymore.

Meaning that you need to check for all item that could have been green, you need to check all the items that can be colored. On top of that, if "Tshirt" doesn't have the attribute "Color" anymore, it's (i believe) impossible to get to "Tshirt" from "Green" (or from "Color") and thus you would need to cycle on each product and check if any of their templateAttributeValue is linked to the AttributeValue "Green" (Or maybe cycle on each templateAttributeValue directly and then get the product if the value match)

**Comment by Feyensv** [comment]
> Meaning that you need to check for all item that could have been green, you need to check all the items that can be colored. On top of that, if "Tshirt" doesn't have the attribute "Color" anymore, it's (i believe) impossible to get to "Tshirt" from "Green" (or from "Color") and thus you would need to cycle on each product and check if any of their templateAttributeValue is linked to the AttributeValue "Green" (Or maybe cycle on each templateAttributeValue directly and then get the product if the value match)

If there was no link between Green and Tshirt, there wouldn't be an error when you try to delete Green.
The link remaining between the two records is indeed probably a product.template.attribute.value record (Since `product_variant_combination` relation is mentioned in the error message in the video from the support ticket).
Through that link, you should be able to find the products, and verify if they can be deleted.

**Comment by fdamhaut** [comment]
> If there was no link between Green and Tshirt, there wouldn't be an error when you try to delete Green. The link remaining between the two records is indeed probably a product.template.attribute.value record (Since `product_variant_combination` relation is mentioned in the error message in the video from the support ticket). Through that link, you should be able to find the products, and verify if they can be deleted.

The error come from an SQL error when you try to delete an SQL foreign key, which explain why the error return no model
image

**Comment by fdamhaut** [comment]
et `product_variant_combination` c'est une relation entre ProductTemplateAttributeValue et Product mais le probleme c'est que la relation entre ProductTemplateAttributeValue et ProductAttributeValue est à sens unique depuis PTAV vers PAV donc pas moyen d'y avoir acces depuis PAV

**Comment by Feyensv** [comment]
pas par un field, mais par un search, oui.

**Comment by fdamhaut** [comment]
Donc actuellement le chemin le plus court c'est 

-> ProductAttrVal -search> List[ProdTemplAttrVal] -map> List[ProdTempl] -map*> variants_ids

Le probleme c'est que le map* utilise `_get_possible_variants `

Qui dit : 

         If there are a lot of variants, this method might be slow. Even if there
        aren't too many variants, for performance reasons, do not call this
        method in a loop over the product templates.

        Therefore this method has a very restricted reasonable use case and you
        should strongly consider doing things differently if you consider using
        this method.

...

**Comment by fdamhaut** [comment]
AHHHHHHHHHH Pardon je suis aveugle j'étais passé au dessus de `ptav_product_variant_ids` sans le voir. Du coup, Merci ^^

File: addons/product/models/product_attribute.py [hunk 1]
```diff
@@ -148,9 +148,13 @@ def unlink(self):
...
             if pav.is_used_on_products:
                 raise UserError(
-                    _("You cannot delete the value %s because it is used on the following products:\n%s") %
+                    _("You cannot delete the value %s because it is used on the following products:\n%s\nIf the value has been associated to other product in the past, you will not be able to deleted it.") %
                     (pav.display_name, ", ".join(pav.pav_attribute_line_ids.product_tmpl_id.mapped('display_name')))
                 )
+            linked_products = pav.env['product.template.attribute.value'].search([('product_attribute_value_id', '=', pav.id)]).with_context(active_test=False).ptav_product_variant_ids
+            unlinkable_products = linked_products._filter_to_unlink()
+            if linked_products != unlinkable_products:
+                raise UserError(_("Cannot delete value %s because it was used in some products", pav.display_name))
```
Comment (Feyensv) [inline addons/product/models/product_attribute.py:157]: This is a new translated string, you have to update the .pot files when you modify translated strings in stable version (to request new translations on Transifex for the new/modified string, and create corresponding records in database).

Comment (fdamhaut) [reply addons/product/models/product_attribute.py]: :ok_hand:

**Comment by fdamhaut** [comment]
@robodoo r+

**Comment by Feyensv** [comment]
You should request a mail update in mergebot backend from the runbot team (it's due to a recent change in the mergebot).

**Comment by fdamhaut** [comment]
@robodoo r+ 
please
